packages/dds/tree/src/test/utils.ts

```diff
@@ -83,11 +84,14 @@ export class TestTreeProvider {
      * await trees.ensureSynchronized();
      * ```
      */
-    public static async create(trees = 0): Promise<ITestTreeProvider> {
+    public static async create(trees = 0, summarizeOnDemand = false): Promise<ITestTreeProvider> {
         const provider = new TestTreeProvider() as ITestTreeProvider;
         for (let i = 0; i < trees; i++) {
```

Comment:
I think the optimal logic here is:
1. Assert trees !== 0 || summarizeOnDemand is false
2. Create the first tree if trees > 0
3. Create the summarizer if summarizeOnDemand is true. Save the summarizer in a field
4. Create the rest of the trees

This implies replacing the summarizeOnDemand field with the summarizer itself. Then the manualSummarize() method should assert that this.summarizer !== undefined before summarizing.

Reply:
Sounds good Noah :)

However, if we create the summarizer within the for loop (immediately after the first tree), will we still need the assert statement?
```
    public static async create(trees = 0, summarizeOnDemand = false): Promise<ITestTreeProvider> {
        const provider = new TestTreeProvider() as ITestTreeProvider;
        for (let i = 0; i < trees; i++) {
            await provider.createTree();
            if (summarizeOnDemand && i === 0) {
                const summarizer = await createSummarizer(provider, provider.containers[0]);
                this.summarizer = async () => {
                    await summarizeNow(summarizer, "TestTreeProvider");
                };
            }
        }

        return provider;
    }
```

Reply:
Yes. The way you wrote it above is elegant (nice work!) and it does indeed skip the problematic case. However, I'm of the opinion that we should prefer throwing errors over silently "failing". For example, right now, if a user calls `summarize()`, and the summarizer is undefined, we do nothing. It would be better for a user if we instead right away threw an error that said "can't summarize because... " (this is _especially_ true because this is a test utility. Sometimes in production code it's preferable to fail silently rather than throwing errors so that a user doesn't end up seeing an error message, but in test helper code we can be aggressive about emitting errors because we want to be sure that the developer is using the utilities correctly!). So, what do we put after the "because..."? Well, `summarize` could either fail because 'summarizeOnDemand' was never set to true (oops! Let's tell the dev to set it to true!). Or, it could fail because, although they set `summarizeOnDemand` to true, they set `trees` to 0. In that case, we want to give them a different error message ("must create at least one tree"). We can't tell inside the `summarize` method which of these two cases we're in (the summarizer is undefined but we don't know why), but if we instead move the second error to happen right away in `create()`, we dodge that problem. And anyway, it is usually best to error as soon as possible so that the dev can fix their code a quickly as possible. We know that if they call `create(0, true)` then they must not understand the contract. Let's educate them right away with an error rather than letting it slide and letting them remain confused :)

Reply:
Two assert statements in `create` and `summarize` methods added.
create:
```
        assert(
            !(trees === 0 && summarizeOnDemand),
            "can't summarize, because must create at least one tree.",
        );
```
summarizer:
```
        assert(
            this.summarizer !== undefined,
            "can't summarize because summarizeOnDemand was not set to true.",
        );
```